Concrete projection of a VPolytope

... should return a VPolytope

```julia
X = rand(VPolytope, dim=3)
project(X, [2])

AssertionError: package 'Polyhedra' not loaded (it is required for executing `default_polyhedra_backend`)

Stacktrace:
 [1] macro expansion at /home/mforets/.julia/dev/LazySets/src/Assertions/Assertions.jl:23 [inlined]
 [2] require(::Symbol; fun_name::String, explanation::String) at /home/mforets/.julia/dev/LazySets/src/init.jl:37
 [3] default_polyhedra_backend(::VPolytope{Float64,Array{Float64,1}}, ::Type{T} where T) at /home/mforets/.julia/dev/LazySets/src/Interfaces/AbstractPolyhedron_functions.jl:24
 [4] tohrep(::VPolytope{Float64,Array{Float64,1}}) at /home/mforets/.julia/dev/LazySets/src/Sets/VPolytope.jl:518
 [5] constraints_list(::VPolytope{Float64,Array{Float64,1}}) at /home/mforets/.julia/dev/LazySets/src/Sets/VPolytope.jl:378
 [6] constrained_dimensions(::VPolytope{Float64,Array{Float64,1}}) at /home/mforets/.julia/dev/LazySets/src/Interfaces/AbstractPolyhedron_functions.jl:120
 [7] project(::VPolytope{Float64,Array{Float64,1}}, ::Array{Int64,1}) at /home/mforets/.julia/dev/LazySets/src/Approximations/decompositions.jl:638
 [8] top-level scope at In[22]:2
```

--- a/src/Approximations/decompositions.jl
+++ b/src/Approximations/decompositions.jl
@@ -665,5 +665,5 @@ function project(V::Union{<:VPolygon{N}, <:VPolytope{N}},
     n = dim(V)
     M = projection_matrix(block, n, N)
     πvertices = broadcast(v -> M * v, vertices_list(V))
-    return basetype(V)(πvertices)
+    return VPolytope(πvertices)
 end

--- a/src/Approximations/decompositions.jl
+++ b/src/Approximations/decompositions.jl
@@ -659,3 +659,11 @@ function project(H::AbstractHyperrectangle, block::AbstractVector{Int})
     πr = radius_hyperrectangle(H)[block]
     return Hyperrectangle(πc, πr, check_bounds=false)
 end
+
+function project(V::Union{<:VPolygon{N}, <:VPolytope{N}},
+                 block::AbstractVector{Int}) where {N}
+    n = dim(V)
+    M = projection_matrix(block, n, N)
+    πvertices = broadcast(v -> M * v, vertices_list(V))
+    return basetype(V)(πvertices)
+end

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -391,6 +391,10 @@ for N in [Float64, Float32, Rational{Int}]
     m = N[4 0; 6 2; 4 4]'
     P = VPolygon(m)
     @test is_cyclic_permutation(vertices_list(P), [N[4, 0], N[6, 2], N[4, 4]])
+
+    # test concrete projection
+    V = VPolygon([N[0, 1], N[1, 0], N[-1, 0]])
+    @test project(V, [1]) == VPolygon([N[-1], N[1]])
 end
 
 function same_constraints(v::Vector{<:LinearConstraint{N}})::Bool where N<:Real

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -395,6 +395,8 @@ for N in [Float64, Float32, Rational{Int}]
     # test concrete projection
     V = VPolygon([N[0, 1], N[1, 0], N[-1, 0]])
     @test project(V, [1]) == VPolygon([N[-1], N[1]])
+    V = VPolygon([N[1, 0], N[1, 1]])
+    @test project(V, [1]) == VPolygon([N[1], N[1]])
 end
 
 function same_constraints(v::Vector{<:LinearConstraint{N}})::Bool where N<:Real

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -394,7 +394,7 @@ for N in [Float64, Float32, Rational{Int}]
 
     # test concrete projection
     V = VPolygon([N[0, 1], N[1, 0], N[-1, 0]])
-    @test project(V, [1]) == VPolytope([N[-1], N[1]])
+    @test project(V, [1]) == VPolytope([N[-1], N[1], N[0]])
     V = VPolygon([N[1, 0], N[1, 1]])
     @test project(V, [1]) == VPolytope([N[1], N[1]])
 end

--- a/test/unit_Polygon.jl
+++ b/test/unit_Polygon.jl
@@ -394,9 +394,9 @@ for N in [Float64, Float32, Rational{Int}]
 
     # test concrete projection
     V = VPolygon([N[0, 1], N[1, 0], N[-1, 0]])
-    @test project(V, [1]) == VPolygon([N[-1], N[1]])
+    @test project(V, [1]) == VPolytope([N[-1], N[1]])
     V = VPolygon([N[1, 0], N[1, 1]])
-    @test project(V, [1]) == VPolygon([N[1], N[1]])
+    @test project(V, [1]) == VPolytope([N[1], N[1]])
 end
 
 function same_constraints(v::Vector{<:LinearConstraint{N}})::Bool where N<:Real

--- a/test/unit_Polytope.jl
+++ b/test/unit_Polytope.jl
@@ -194,6 +194,11 @@ for N in [Float64, Rational{Int}, Float32]
     p[1] = N(5)
     # test that Pcopy is independent of P ( = deepcopy)
     @test Pcopy.vertices[1] == [N(1)]
+
+    # test concrete projection
+    V = VPolytope([N[0, 0, 1], N[0, 1, 0], N[0, -1, 0], N[1, 0, 0]])
+    @test project(V, [1]) == VPolytope([N[0], N[0], N[0], N[1]])
+    @test project(V, [1, 2]) == VPolytope([N[0, 0], N[0, 1], N[0, -1], N[1, 0]])
 end
 
 # default Float64 constructors

